### ISSUE TITLE
Instructor Mode with GPT

### ISSUE THREAD

**Issue by justinian336** [freeCodeCamp/freeCodeCamp#49759]
### Is your feature request related to a problem? Please describe.

No, it's about adding a new feature

### Describe the solution you'd like

I'd like to see an instructor chatbot. Basically it could be a small chat window on the lower corner of the screen of all the exercises.
The chatbot would be able to talk with the student, give useful hints in case the student gets stuck, encourage the student to not give up easily, and explain more in depth the concepts that are necessary to solve the exercise.
GPT has multilingual capabilities, so it could chat with students in their native language.

### Describe alternatives you've considered

I have access to GPT-4 and have tried some prompts in the OpenAI playground that achieve a similar effect to what I'm describing. The `gpt-3.5-turbo` model is way cheaper and could be good enough. Also, there are some open source alternatives in development that in the medium term could become viable replacements at a lower cost.

### Additional context

Khan Academy will start doing something similar and it looks very useful.
https://blog.khanacademy.org/harnessing-ai-so-that-all-students-benefit-a-nonprofit-approach-for-equal-access/

**Comment by Sboonny** [comment]
If khanacademy plan to teach students using chatbots with less than 70% accuracy, I don't know why we should follow the same route, if they want to teach something that can be wrong, they are free to, but I don't see a reason to follow their steps.

**Comment by justinian336** [comment]
@Sboonny 70% on what benchmark? I understand your point that LLMs are prone to hallucinations, so it would be necessary to assess whether it is possible to prevent that from having an impact on the experience of the students. Obviously it would be necessary to do PDCA before incorporating a chatbot officially into the platform.

> I don't know why we should follow the same route.

I think a feature like that could improve the engagement of students. For example, students would be more likely to make easy questions without fearing "sounding dumb" to a chatbot than to other people in the community. Also, it would improve the experience for non-English speaking persons, because sometimes it can be hard to find supportive people in your own language. It is not a substitute of the community, but it can make the platform more fun and inclusive.

I just believe that it would be interesting to explore the possibility.

**Comment by Sboonny** [comment]
From OpenAI research paper, https://arxiv.org/pdf/2303.08774.pdf

At "Internal factual eval by category" section, you notice that code isn't above 70%, you are free to discuss it, I haven't closed the issue and this is personal opinion, these tools aren't ready to teach code if they are confidentially lie more than 30% of the times.

**Comment by ojeytonwilliams** [comment]
I think a tool like this could be a great addition to freeCodeCamp, in principle. However, I share @Sboonny's concern about the failure rate and the way that it fails.

Here's how I would categorize the behaviour (ignoring safety for a moment)

|       | Answers | Refuses |
| ----------- | ----------- | -- |
| **True**      |  Helpful  | Excessively cautious |
| **False**   | Misleading        | Justifiably cautious |

To my mind, of those four, we should focus primarily on when it is misleading, since that is far, far worse than the other three. If it is possible reduce the chance that it produces misleading answers enough, then it is potentially useful. I think that is true even if the helpful rate is low, since we can always direct people to the forum if the bot cannot help.

Finally, it's not sufficient to stop it being misleading, it has to be safe. For instance, it should refuse to answer questions that are not directly related to the topics we teach.

If it's possible to produce a safe, reliable and moderately useful tutor, then that's great. However, I don't know enough about the technology to know if that's possible at the moment.

**Comment by ShaunSHamilton** [comment]
For the most part, I like the idea of a programming chat assistant.

My reservations for adding one to freeCodeCamp are:
- Cost
  - Existing solutions are completely beyond the donor budget (the price per token sounds low, but after adding the number of Campers, it becomes unsustainable)
  - We are not at a point where we could handle the increase in requests (server compute) without spending too much on more server-space.
- As mentioned, we would want a bot that only knows about programming concepts
  - One such model might not exist yet

With that on mind, we circle back to the fact that Campers can go to search engines and existing chatbots themselves (in a separate browser tab). _With the obvious pitfall of missing context_, but, with the argument that a chatbot is meant for extended conversation (not just quick "Google"), then adding the context manually is not a waste of time. _just thoughts_

**Comment by justinian336** [comment]
@Sboonny I think that the factuality evaluations provided in the paper may not reflect the actual performance of GPT when used on the FreeCodeCamp codebase. First, because it's not clear that we would be forced to use the same generation settings they used in their evaluation. For example, if you set the temperature too high you might see more hallucinations, because the model literally includes randomness in the next-token prediction.

Also, it is likely that the FreeCodeCamp exercises are already on GPT's training set. Also, the paper recommends *grounding* the model to avoid factual mistakes, which is possible to do given that the information of the solution of the problems, the hints, etc. can be made available to the model. I think that the disadvantages you mention cannot be avoided, but can be mitigated to a certain extent. I guess it's only possible to know the cost-benefit balance through experimentation.

**Comment by jeremylt** [comment]
We've seen that GitHub's Copilot will happily regurgitate entire challenge solutions for users. How do we prevent users from effectively doing an expensive copy-paste of the answer on fCC's dime?

**Comment by Sboonny** [comment]
Although the instructor editor is part of the future, and we would welcome it, but it isn't currently ready for the current curriculum and it's may be expensive with the limit money we get from the doners. 

We are happy to re discuss this in the future.
